pkg/tfshim/walk/walk.go

```diff
@@ -0,0 +1,198 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package walk
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/schema"
+)
+
+// Represents locations in a Schema value as a sequence of steps to locate it.
+//
+// An empty SchemaPath represents the current location.
+//
+// Values of this type are immutable by convention, use Copy as necessary for local mutations.
+type SchemaPath []SchemaPathStep
+
+func (p SchemaPath) String() string {
+	parts := []string{"walk", "NewSchemaPath()"}
+	for _, e := range p {
+		switch ee := e.(type) {
+		case ElementStep:
+			parts = append(parts, "Element()")
+		case GetAttrStep:
+			parts = append(parts, fmt.Sprintf("GetAttr(%q)", ee.Name))
+		}
+	}
+	return strings.Join(parts, ".")
+}
+
+func (p SchemaPath) Copy() SchemaPath {
+	ret := make(SchemaPath, len(p))
+	copy(ret, p)
+	return ret
+}
+
+func (p SchemaPath) Element() SchemaPath {
+	return p.WithStep(ElementStep{})
+}
+
+func (p SchemaPath) GetAttr(name string) SchemaPath {
+	return p.WithStep(GetAttrStep{name})
+}
+
+func (p SchemaPath) WithStep(suffix SchemaPathStep) SchemaPath {
+	ret := make(SchemaPath, len(p)+1)
+	copy(ret, p)
+	ret[len(p)] = suffix
+	return ret
+}
+
+// Builds a new empty SchemaPath.
+func NewSchemaPath() SchemaPath {
+	return make(SchemaPath, 0)
+}
+
+// Finds a nested Schema at a given path.
+func LookupSchemaPath(path SchemaPath, schema shim.Schema) (shim.Schema, error) {
+	p := path
+	current := NewSchemaPath()
+	result := schema
+	for {
+		if len(p) == 0 {
+			return result, nil
+		}
+		nextResult, err := p[0].Lookup(result)
+		if err != nil {
+			return nil, fmt.Errorf("LookupSchemaPath failed at %s: %w", current, err)
+		}
+		result, p, current = nextResult, p[1:], current.WithStep(p[0])
+	}
+}
+
+// Similar to LookupSchemaPath but starts the initial step from a SchemaMap.
+func LookupSchemaMapPath(path SchemaPath, schemaMap shim.SchemaMap) (shim.Schema, error) {
+	return LookupSchemaPath(path, wrapSchemaMap(schemaMap))
+}
+
+// Represents elements of a SchemaPath.
+//
+// This interface is closed, the only the implementations given in the current package are allowed.
+type SchemaPathStep interface {
+	String() string
+	Lookup(shim.Schema) (shim.Schema, error)
+}
+
+// Drill down into an attribute by the given attribute name.
+type GetAttrStep struct {
+	Name string
+}
+
+func (step GetAttrStep) String() string {
+	return fmt.Sprintf("walk.GetAttrStep{%q}", step.Name)
+}
+
+func (step GetAttrStep) Lookup(s shim.Schema) (shim.Schema, error) {
+	if sm, ok := UnwrapSchemaMap(s); ok {
+		s, found := sm.GetOk(step.Name)
+		if !found {
+			return nil, fmt.Errorf("%s not found", step.String())
+		}
+		return s, nil
+	}
+	return nil, fmt.Errorf("%s is not applicable", step.String())
+}
+
+// Drill down into a Map, Set or List element schema.
+type ElementStep struct{}
```

Comment:
So this is a great question. By resource I understand you mean shim.Resource which actually is our current mis-named tfshim encoding of an Object type schema. This is the bit of the design I went back and forth on a bit as it wasn't super clear what to do.

Currently what the code does is this, if you want to find an object type, you search for it's location and you get back a shim.Schema with s.Type()=TypeMap and s.Elem()=shim.Resource with s.Elem().Schema() giving you a SchemaMap describing the resource's properties.  Although this encoding is using Elem(), the path will not us ElementPath{} because actual values of this type will typically not be drilling down through `x[y]` accessors.

This is elaborated in the tests a bit, one sec.